wagtail/core/models/__init__.py

```diff
@@ -1547,6 +1561,16 @@ def move(self, target, pos=None, user=None):
         )
         logger.info("Page moved: \"%s\" id=%d path=%s", self.title, self.id, new_url_path)
 
+        if url_path_changed:
```

Comment:
All the existing methods on page hide away the ``url_path`` and express page URLs with a site/path combination. So I wonder if we should implement this signal handler to be site-aware? I think this will simplify the logic in ``autocreate_redirects`` function too (no need to hack the ``url_path`` field back to its old value to find the old path, since the old path itself would be to hand).

Do you think we could rename this to ``page_url_changed`` and replace the ``url_paths`` with ``site/path`` instead? If the page is visible on multiple sites, we could call this signal handler multiple times.

So, for example, instead of:

```python
page_url_path_changed(instance=page, url_path_before="/home/about/", url_path_after="/home/renamed/")
```

Could we do:

```python
page_url_changed(instance=page, site_before=site, path_before="/about/", site_after=site, path_after="/renamed/")
```

Reply:
Thinking about where the complexity lives is definitely useful, but I feel this suggestion gets it wrong in a couple of ways:
1. It involves complicating the Page.save() and Page.move() methods further in order to have all of those values to hand, for functionality that is entirely optional. The current solution is lightweight in terms of changes to `Page`, because `page_url_path_changed()` uses values that are mostly already available.
2. It assumes a lot more about how URLs work within a project. 

While we should not complicate things unnecessarily for every edge-case we can think of, if having a little more complication in the (optional) signal handler means that we keep everything else simple/generic, I would say that's probably the better way to go?

Reply:
> It involves complicating the Page.save() and Page.move() methods further in order to have all of those values to hand, for functionality that is entirely optional. The current solution is lightweight in terms of changes to Page, because page_url_path_changed() uses values that are mostly already available.

The current method probably makes sense for automatic redirects, but if this is meant to become part of the public API for pages, I feel it's important to be consistent with the existing methods.

I'd think we'd implement this as a method or a function so it wouldn't have to make the implementations of those functions any more complicated than they would be if they were using ``url_path``?

Alternatively, could we implement a signal for changing the slug instead? And use the ``page_moved`` signal?

Reply:
> The current method probably makes sense for automatic redirects, but if this is meant to become part of the public API for pages, I feel it's important to be consistent with the existing methods.

Okay, I see the importance of that too. Maybe there's some meeting in the middle to be had - I'm just not seeing it yet.

> Alternatively, could we implement a signal for changing the slug instead? And use the `page_moved` signal?

I'd be happy to go with that, but the `url_path_before` and `url_path_after` values are really crucial to the current signal handler solution, and I'm not sure they would be a good fit for a `page_slug_changed` signal.

If we can figure out an alternative way to retrieve all of the old page URLs without the `url_path_before` and `url_path_after` values (and the current in-memory change trick), I'd be more than happy using `page_slug_changed` and `post_page_move`.

Reply:
@kaedroho  `page_url_path_changed` is now gone, and in it's place is:

```
page_slug_changed(instance=page, instance_before=old_page)
```

We're not claiming that the url has changed, so isn't misleading. And it's useful to have the old and new versions to hand, because it allows you to see ALL of the fields that changed (not just the `slug`).